## **Section 3:** Geospatial, finally
##### The names are all standardized/ normalized, now we can get to the geospatial
##### Through trial and error I've learned that we need to buffer the geometries, to overcome minor issues
##### And that buffering in 4326 avoids an issue with the geometry/boundary of Fiji, which has a landmass that crosses the 180/0 latitude

In [1]:
# Import a few packages
import pandas as pd
import geopandas as gpd
import matplotlib.pyplot as plt
import folium
import os
pd.options.display.max_rows = 300

In [2]:
print(os.getcwd())
os.chdir('G:/My Drive/Clark')
print(os.getcwd())

G:\My Drive\Clark\GIS Tutorials\Geog-312\Geog-312\1.Geopandas
G:\My Drive\Clark


In [3]:
# And bring in the edited country boundary shapefile
# These are very detaile and precise polygons. Takes a few moments to load
countries_w_off_path = "GIS Tutorials/Geog-312/geopandas_Files/checkpoint2/countries_w_off_names.shp"
countries_w_off_names = gpd.read_file(countries_w_off_path)
print(len(countries_w_off_names))
countries_w_off_names.head(10)

248


COUNTRY   ISO           COUNTRYAFF  Population  \
0            Afghanistan    AF          Afghanistan    34262840   
1  Akrotiri and Dhekelia  None       United Kingdom       18195   
2                Albania    AL              Albania     2402113   
3                Algeria    DZ              Algeria    46700000   
4         American Samoa    AS        United States       49710   
5                Andorra    AD              Andorra       86398   
6                 Angola    AO               Angola    35121734   
7               Anguilla    AI       United Kingdom       15780   
8             Antarctica    AQ                 None           0   
9    Antigua and Barbuda    AG  Antigua and Barbuda      103603   

                                formalName  \
0           Islamic Emirate of Afghanistan   
1                                     None   
2                      Republic of Albania   
3  People's Democratic Republic of Algeria   
4                                     None   
5                  Principality of Andorra   
6                       Republic of Angola   
7                                     None   
8                                     None   
9                      Antigua and Barbuda   

                                            geometry  
0  POLYGON ((74.88986 37.23409, 74.88962 37.23314...  
1  MULTIPOLYGON (((32.8388 34.70555, 32.84127 34....  
2  MULTIPOLYGON (((20.0789 42.5558, 20.07939 42.5...  
3  MULTIPOLYGON (((8.64188 36.94206, 8.64196 36.9...  
4  MULTIPOLYGON (((-171.07753 -11.06622, -171.080...  
5  POLYGON ((1.7258 42.5044, 1.71149 42.49224, 1....  
6  MULTIPOLYGON (((13.10288 -4.68421, 13.10173 -4...  
7  MULTIPOLYGON (((-63.42216 18.59739, -63.42672 ...  
8  MULTIPOLYGON (((-46.15775 -60.51078, -46.1787 ...  
9  MULTIPOLYGON (((-61.84592 17.72958, -61.83383 ...

In [4]:
# Verify CRS and buffer. Takes a few moments to run
print(countries_w_off_names.crs)
countries_buffed = countries_w_off_names
countries_buffed['geometry'] = countries_buffed['geometry'].buffer(0)

EPSG:4326


In [5]:
# # Plot county borders
# # They are very detailed boundary polygons and this takes a littlewhile to render/plot
# # REMEMBER to clear output of this and all folium map cells after inspecting
# # Leaving them open in a jupyter cell uses a lot of memory
# # Other option is to save an HTML of the folium map and inspect in browser window

# import folium
# m = folium.Map(location=[0,0], zoom_start=2)

# # Add only the polygon geometry as GeoJSON (avoiding LineStrings or other non-polygon geometries)
# folium.GeoJson(
#     countries_w_off_names[['geometry']],  # Use only the polygons from 'geometry' column
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue'},
#     name="Country Boundaries"
# ).add_to(m)

# # OPTIONAL: Save the map
# # m.save('GIS Tutorials/GeoPy/1.Geopandas/activity_files/folium_maps/countries_boundary_check.html')

# # Or show the map
# m

##### No need to run the two cells below
##### But if you do, the result of buffering in EPSG:6933 is the boundary of Fiji comes to include a few glitchy lines that run all the way around the equator!

In [6]:
# # Now, to compare, do the buffering in 6933
# buff6933 = countries_w_off_names.copy()
# buff6933 = buff6933.to_crs("EPSG:6933")
# buff6933['geometry'] = buff6933['geometry'].buffer(0)

In [7]:
# # # Plot county borders with glitchy Fiji geometry
# import folium
# m = folium.Map(location=[0,0], zoom_start=2)

# # Add only the polygon geometry as GeoJSON (avoiding LineStrings or other non-polygon geometries)
# folium.GeoJson(
#     buff6933[['geometry']],  # Use only the polygons from 'geometry' column
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue'},
#     name="Country Boundaries"
# ).add_to(m)

# # Save weird map
# m.save('GIS Tutorials/GeoPy/1.Geopandas/activity_files/folium_maps/fiji_glitch.html')

# # Show the map
# m

In [8]:
# # Or, faster, create/save map with JUST glitchy vs good fiji, to compare

# # Initialize a Folium map
# m = folium.Map(location=[-17.7134, 178.0650], zoom_start=5)  # Centering near Fiji's approximate location

# # Filter the GeoDataFrame to include only Fiji's boundary
# fiji_good = countries_w_off_names[countries_w_off_names['COUNTRY'] == 'Fiji']

# # Add Fiji's boundary to the map as GeoJSON
# folium.GeoJson(
#     fiji_good[['geometry']],  # Only Fiji's geometry
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'red', 'weight': 2},
#     name="Good Fiji"
# ).add_to(m)

# # Filter the GeoDataFrame to include only Fiji's boundary
# glitchyFiji = buff6933[buff6933['COUNTRY'] == 'Fiji']

# # Add Fiji's boundary to the map as GeoJSON
# folium.GeoJson(
#     glitchyFiji[['geometry']],  # Only Fiji's geometry
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue', 'weight': 2},
#     name="Glitch Fiji"
# ).add_to(m)

# # Add a LayerControl to toggle layers on and off
# folium.LayerControl().add_to(m)

# # # Save the map
# # m.save('GIS Tutorials/GeoPy/1.Geopandas/activity_files/folium_maps/fji_boundary_compare.html')

# # Display the map if running in an interactive notebook
# m

##### Buffering is best done in 4326, but calculating centroids, which requires accurate area, should be done in our projected coordinate system, 6933

In [9]:
print(countries_buffed.crs)
# Reproject to a global equal-area projection
# This is necessary when doing any kind of distance calculation, such as determining centroids
countries6933 = countries_buffed.to_crs("EPSG:6933")
print(countries6933.crs)
# Centroids should be calculated in the projected CRS
countries6933['centroid'] = countries6933.geometry.centroid

EPSG:4326
EPSG:6933


In [10]:
# Verify that centroids and geometry are both in meters
countries6933.head()

COUNTRY   ISO      COUNTRYAFF  Population  \
0            Afghanistan    AF     Afghanistan    34262840   
1  Akrotiri and Dhekelia  None  United Kingdom       18195   
2                Albania    AL         Albania     2402113   
3                Algeria    DZ         Algeria    46700000   
4         American Samoa    AS   United States       49710   

                                formalName  \
0           Islamic Emirate of Afghanistan   
1                                     None   
2                      Republic of Albania   
3  People's Democratic Republic of Algeria   
4                                     None   

                                            geometry  \
0  POLYGON ((7225844.213 4429969.158, 7225820.477...   
1  MULTIPOLYGON (((3271001.456 4208757.218, 32710...   
2  MULTIPOLYGON (((1937337.89 4953531.35, 1937385...   
3  MULTIPOLYGON (((833823.241 4400105.218, 833831...   
4  MULTIPOLYGON (((-16226341.969 -1837186.443, -1...   

                            centroid  
0    POINT (6366923.369 4067003.428)  
1    POINT (3218981.275 4180831.548)  
2    POINT (1936226.181 4816925.087)  
3     POINT (258594.776 3423503.594)  
4  POINT (-16440866.53 -1795748.572)

In [11]:
# But for science, what happens if we calculate in 4326?
# Notice the warning
countries_buffed['centroid'] = countries_buffed.geometry.centroid

C:\Users\User\AppData\Local\Temp\ipykernel_26516\2017718633.py:3: UserWarning: Geometry is in a geographic CRS. Results from 'centroid' are likely incorrect. Use 'GeoSeries.to_crs()' to re-project geometries to a projected CRS before this operation.

  countries_buffed['centroid'] = countries_buffed.geometry.centroid


In [12]:
def reproject_geom_and_centroid(gdf, target_crs):
    """
    Reprojects the given GeoDataFrame to the target CRS, calculates the centroid coordinates
    in the target CRS, and adds the centroid coordinates as new columns to the reprojected GDF.

    Parameters:
    gdf (GeoDataFrame): The input GeoDataFrame with geometries to reproject.
    target_crs (str or int): The target CRS, either as an EPSG code or a PROJ string.

    Returns:
    GeoDataFrame: The reprojected GeoDataFrame with centroid coordinates as new columns.
    """
    # Reproject the GeoDataFrame to the target CRS
    gdf_reproj = gdf
    gdf_reproj = gdf_reproj.to_crs(epsg=target_crs)
    gdf_reproj['centroid'] = gdf_reproj['centroid'].to_crs(epsg=target_crs)

    # Return the reprojected GeoDataFrame with centroid coordinates
    return gdf_reproj

In [13]:
countries6933_reproj = reproject_geom_and_centroid(countries6933, '4326')
print("Check that countries4326 geometry and centroid are both in lat/lon coords:")
print(countries6933_reproj.crs)
countries6933_reproj.head()

Check that countries4326 geometry and centroid are both in lat/lon coords:
EPSG:4326


COUNTRY   ISO      COUNTRYAFF  Population  \
0            Afghanistan    AF     Afghanistan    34262840   
1  Akrotiri and Dhekelia  None  United Kingdom       18195   
2                Albania    AL         Albania     2402113   
3                Algeria    DZ         Algeria    46700000   
4         American Samoa    AS   United States       49710   

                                formalName  \
0           Islamic Emirate of Afghanistan   
1                                     None   
2                      Republic of Albania   
3  People's Democratic Republic of Algeria   
4                                     None   

                                            geometry  \
0  POLYGON ((74.88986 37.23409, 74.88962 37.23314...   
1  MULTIPOLYGON (((33.90121 35.09612, 33.90185 35...   
2  MULTIPOLYGON (((20.0789 42.5558, 20.07939 42.5...   
3  MULTIPOLYGON (((8.64188 36.94206, 8.64196 36.9...   
4  MULTIPOLYGON (((-168.17253 -14.55294, -168.173...   

                      centroid  
0    POINT (65.98786 33.75477)  
1    POINT (33.36206 34.83019)  
2    POINT (20.06737 41.12698)  
3     POINT (2.68012 27.89881)  
4  POINT (-170.3959 -14.21789)

In [14]:
# # But there is still an issue with the centroid of Fiji!
# # Plot centroids in each projection, and the country boundary of Fiji
# # Interesting to note how the difference in 4326 and 6933 centroid calculation is greater when further from the equator
# # Create the map
# m = folium.Map(location=[0, 0], zoom_start=2)

# # Filter the GeoDataFrame to include only Fiji's boundary
# fiji_boundary = countries_buffed[countries_buffed['COUNTRY'] == 'Fiji']

# # Add only Fiji's boundary to the map as a layer
# fiji_boundary_layer = folium.FeatureGroup(name="Fiji Boundary")
# folium.GeoJson(
#     fiji_boundary[['geometry']],  # Use only Fiji's geometry
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue', 'weight': 2}
# ).add_to(fiji_boundary_layer)
# fiji_boundary_layer.add_to(m)

# # Add the centroids as calculated in the geographic CRS (4326) as circle markers, with Fiji in a different color
# centroids_4326_layer = folium.FeatureGroup(name="Centroids 4326")
# for _, row in countries_buffed.iterrows():
#     centroid_color = 'yellow' if row['COUNTRY'] == 'Fiji' else 'blue'  # Fiji in yellow, others in blue
#     folium.CircleMarker(
#         location=[row['centroid'].y, row['centroid'].x],
#         radius=3,
#         color=centroid_color,
#         fill=True,
#         fill_color=centroid_color,
#         fill_opacity=1 
#     ).add_child(folium.Tooltip(f"Geographic CRS (4326) centroid: {row['COUNTRY']}")).add_to(centroids_4326_layer)
# centroids_4326_layer.add_to(m)

# # Add the centroids as calculated the projected CRS (6933) as circle markers, with Fiji in a different color
# centroids_6933_layer = folium.FeatureGroup(name="Centroids 6933")
# for _, row in countries6933_reproj.iterrows():
#     centroid_color = 'red' if row['COUNTRY'] == 'Fiji' else 'green'  # Fiji in red, others in green
#     folium.CircleMarker(
#         location=[row['centroid'].y, row['centroid'].x],
#         radius=3,
#         color=centroid_color,
#         fill=True,
#         fill_color=centroid_color,
#         fill_opacity=1 
#     ).add_child(folium.Tooltip(f"Projected CRS (6933) centroid: {row['COUNTRY']}")).add_to(centroids_6933_layer)
# centroids_6933_layer.add_to(m)

# # Add a LayerControl to toggle layers on and off
# folium.LayerControl().add_to(m)

# # Save map
# m.save('GIS Tutorials/GeoPy/1.Geopandas/activity_files/folium_maps/centroidComparison.html')

# # Show the map
# m

In [15]:
# So, we need to calculate the centroid of Fiji using a more Fiji-specific CRS
# Pull out Fiji boundary from unmodified gdf (no centroids)
fiji_gdf = countries_w_off_names[countries_w_off_names['COUNTRY'] == 'Fiji']
# We will use EPSG:3832 (South Pacific 1958 / Fiji 1958)
fiji_gdf_3832 = fiji_gdf.to_crs(epsg=3832)

# Calculate the centroid of Fiji in the projected CRS
fiji_gdf_3832['centroid'] = fiji_gdf_3832.geometry.centroid

fiji_centroid_3832 = fiji_gdf_3832.geometry.centroid.iloc[0]

# Display the centroid coordinates. Will be in meters
print(f"Fiji centroid in EPSG:3832: {fiji_centroid_3832}")

Fiji centroid in EPSG:3832: POINT (3176924.386078355 -1960148.7447583182)


In [16]:
# # Now project this point gdf to 4326 to check our work
# fiji_reproj_4326 = reproject_and_get_centroid(fiji_gdf_3832, 4326)

# fiji_centroid4326 = fiji_reproj_4326.geometry.centroid.iloc[0]
# # Look what happens if you 
# print(f"Reprojected to 4326: {fiji_centroid4326}")

In [17]:
# Now project this point gdf to 4326 to check our work
fiji_reproj_4326 = reproject_geom_and_centroid(fiji_gdf_3832, 4326)
fiji_reproj_4326

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  MULTIPOLYGON (((179.99667 -16.53378, 179.98753...   

                      centroid  
72  POINT (178.5388 -17.44728)

In [18]:
# Visualize to check new Fiji centroid is correct
m = folium.Map(location=[0, 0], zoom_start=2)

# Filter the GeoDataFrame to include only Fiji's boundary
fiji_boundary = countries_buffed[countries_buffed['COUNTRY'] == 'Fiji']

# Add only Fiji's boundary to the map as a layer
fiji_boundary_layer = folium.FeatureGroup(name="Fiji Boundary")
folium.GeoJson(
    fiji_boundary[['geometry']],  # Use only Fiji's geometry
    style_function=lambda x: {'fillOpacity': 0, 'color': 'blue', 'weight': 2}
).add_to(fiji_boundary_layer)
fiji_boundary_layer.add_to(m)

# Add the centroids as circle markers for the geographic CRS (4326), with Fiji in a different color
centroids_4326_layer = folium.FeatureGroup(name="Centroids 4326")
for _, row in countries_buffed.iterrows():
    centroid_color = 'yellow' if row['COUNTRY'] == 'Fiji' else 'blue'  # Fiji in yellow, others in blue
    folium.CircleMarker(
        location=[row['centroid'].y, row['centroid'].x],
        radius=3,
        color=centroid_color,
        fill=True,
        fill_color=centroid_color,
        fill_opacity=1 
    ).add_child(folium.Tooltip(f"Geographic CRS (4326) centroid: {row['COUNTRY']}")).add_to(centroids_4326_layer)
centroids_4326_layer.add_to(m)

# Add the centroids as circle markers for the projected CRS (6933), with Fiji in a different color
centroids_6933_layer = folium.FeatureGroup(name="Centroids 6933")
for _, row in countries6933_reproj.iterrows():
    centroid_color = 'red' if row['COUNTRY'] == 'Fiji' else 'green'  # Fiji in red, others in green
    folium.CircleMarker(
        location=[row['centroid'].y, row['centroid'].x],
        radius=3,
        color=centroid_color,
        fill=True,
        fill_color=centroid_color,
        fill_opacity=1 
    ).add_child(folium.Tooltip(f"Projected CRS (6933) centroid: {row['COUNTRY']}")).add_to(centroids_6933_layer)
centroids_6933_layer.add_to(m)

# Reproject Fiji's centroid from EPSG:3832 to EPSG:4326 and plot it
fiji_centroid_4326 = fiji_reproj_4326['centroid'].iloc[0]  # Extract the centroid after reprojection
centroid_color = 'pink'  # Fiji centroid in pink
folium.CircleMarker(
    location=[fiji_centroid_4326.y, fiji_centroid_4326.x],  # Use the reprojected centroid
    radius=5,
    color=centroid_color,
    fill=True,
    fill_color=centroid_color,
    fill_opacity=1 
).add_child(folium.Tooltip(f"Fiji Centroid calculated in EPSG:3832")).add_to(m)

# Add a LayerControl to toggle layers on and off
folium.LayerControl().add_to(m)

# # Save
# m.save('GIS Tutorials/GeoPy/1.Geopandas/activity_files/folium_maps/centroidComparison.html')

# Show the map
m

In [19]:
# It worked! 
# Now lets replace the centroid coordinates in countries6933
# Project fiji_gdf_3832 to 6933
fiji_6933= reproject_geom_and_centroid(fiji_gdf_3832, 6933)

fiji_6933

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  MULTIPOLYGON (((17367209.493 -2080868.346, 173...   

                             centroid  
72  POINT (17226544.434 -2192446.141)

In [20]:
# Check that countries 6933 is still in meters for both centroid and geometry
countries6933.head()

COUNTRY   ISO      COUNTRYAFF  Population  \
0            Afghanistan    AF     Afghanistan    34262840   
1  Akrotiri and Dhekelia  None  United Kingdom       18195   
2                Albania    AL         Albania     2402113   
3                Algeria    DZ         Algeria    46700000   
4         American Samoa    AS   United States       49710   

                                formalName  \
0           Islamic Emirate of Afghanistan   
1                                     None   
2                      Republic of Albania   
3  People's Democratic Republic of Algeria   
4                                     None   

                                            geometry  \
0  POLYGON ((7225844.213 4429969.158, 7225820.477...   
1  MULTIPOLYGON (((3271001.456 4208757.218, 32710...   
2  MULTIPOLYGON (((1937337.89 4953531.35, 1937385...   
3  MULTIPOLYGON (((833823.241 4400105.218, 833831...   
4  MULTIPOLYGON (((-16226341.969 -1837186.443, -1...   

                            centroid  
0    POINT (6366923.369 4067003.428)  
1    POINT (3218981.275 4180831.548)  
2    POINT (1936226.181 4816925.087)  
3     POINT (258594.776 3423503.594)  
4  POINT (-16440866.53 -1795748.572)

In [21]:
# Check how wrong the centroid calculation was in this
# Obviously going to be an issue issue with a single piece of land spanning longitude 180/0
countries6933[countries6933['COUNTRY'] == 'Fiji']

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  MULTIPOLYGON (((17367209.493 -2080868.346, 173...   

                           centroid  
72  POINT (279729.909 -2127208.895)

In [22]:
# Pull out the corrected Fiji centroid value
fiji_centroid_6933 = fiji_6933['centroid'].iloc[0]

# Replace the centroid value in countries6933 where 'COUNTRY' is 'Fiji'
countries6933.loc[countries6933['COUNTRY'] == 'Fiji', 'centroid'] = fiji_centroid_6933

# Verify the change
countries6933[countries6933['COUNTRY'] == 'Fiji']

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  MULTIPOLYGON (((17367209.493 -2080868.346, 173...   

                             centroid  
72  POINT (17226544.434 -2192446.141)

In [23]:
# Should be correct centroid coords now
countries6933[countries6933['COUNTRY'] == 'Fiji']

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  MULTIPOLYGON (((17367209.493 -2080868.346, 173...   

                             centroid  
72  POINT (17226544.434 -2192446.141)

In [24]:
# # Lets make sure we added the geometry correctly, and everything is straightened out
# # Reproject back to 4326
# countries_reproj= reproject_geom_and_centroid(countries6933, 4326)

# # And visualize
# m = folium.Map(location=[0, 0], zoom_start=2)

# # Filter the GeoDataFrame to include only Fiji's boundary
# fiji_boundary = countries_buffed[countries_buffed['COUNTRY'] == 'Fiji']

# # Add only Fiji's boundary to the map as a layer
# fiji_boundary_layer = folium.FeatureGroup(name="Fiji Boundary")
# folium.GeoJson(
#     fiji_boundary[['geometry']],  # Use only Fiji's geometry
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue', 'weight': 2}
# ).add_to(fiji_boundary_layer)
# fiji_boundary_layer.add_to(m)

# # Add the centroids as circle markers for the geographic CRS (4326), with Fiji in a different color
# centroids_4326_layer = folium.FeatureGroup(name="Centroids 4326")
# for _, row in countries_buffed.iterrows():
#     centroid_color = 'yellow' if row['COUNTRY'] == 'Fiji' else 'blue'  # Fiji in yellow, others in blue
#     folium.CircleMarker(
#         location=[row['centroid'].y, row['centroid'].x],
#         radius=3,
#         color=centroid_color,
#         fill=True,
#         fill_color=centroid_color,
#         fill_opacity=1 
#     ).add_child(folium.Tooltip(f"Geographic CRS (4326) centroid: {row['COUNTRY']}")).add_to(centroids_4326_layer)
# centroids_4326_layer.add_to(m)

# # Add the centroids as circle markers for the projected CRS (6933), with Fiji in a different color
# centroids_6933_layer = folium.FeatureGroup(name="Centroids 6933")
# for _, row in countries_reproj.iterrows():
#     centroid_color = 'red' if row['COUNTRY'] == 'Fiji' else 'green'  # Fiji in red, others in green
#     folium.CircleMarker(
#         location=[row['centroid'].y, row['centroid'].x],
#         radius=3,
#         color=centroid_color,
#         fill=True,
#         fill_color=centroid_color,
#         fill_opacity=1 
#     ).add_child(folium.Tooltip(f"Projected CRS (6933) centroid: {row['COUNTRY']}")).add_to(centroids_6933_layer)
# centroids_6933_layer.add_to(m)

# # Add a LayerControl to toggle layers on and off
# folium.LayerControl().add_to(m)

# # Show the map
# m

##### After all that, the geometry of fjii is still screwy enough that the centroidOut calculations don't work
##### So here's the deal, Jack
##### We have to run **shapely.make_valid** on the geometries of all the polygons in countries 6933 to calculate the centroids.
##### Fiji's still has to be set manually. 
##### Also, running **make_valid** on the glitchy fiji polygon turns the geometry into a **"GEOMETRYCOLLECTION (MULTIPOLYGON"**
##### So, we will re-replace fiji's geometry from fiji_6933

In [25]:
# # Create field indicating if centroid is in or out of country boundary
# countries_reproj['centroidOut'] = countries_reproj.apply(lambda row: 0 if row['geometry'].contains(row['centroid']) else 1, axis=1)

# print("Check new field, and that countries6933 geometry and centroid are both in meters:")
# # print(countries6933.crs)
# print(countries_reproj.head())

In [26]:
from shapely.validation import make_valid
countries6933_1 = countries6933.copy()
# Create field indicating if centroid is in or out of country boundary
# Apply the centroid containment check only for countries other than Fiji
countries6933_1['geometry'] = countries6933_1['geometry'].apply(lambda geom: make_valid(geom))

countries6933_1['centroidOut'] = countries6933_1.apply(
    lambda row: 0 if row['geometry'].contains(row['centroid']) else 1
    if row['COUNTRY'] != 'Fiji' else None,  # Set Fiji to None temporarily
    axis=1
)

# Set Fiji's 'centroidOut' value manually to 1 (indicating the centroid is out)
countries6933_1.loc[countries6933_1['COUNTRY'] == 'Fiji', 'centroidOut'] = 1

In [27]:
# Should be correct centroid coords now
countries6933_1[countries6933_1['COUNTRY'] == 'Fiji']

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  GEOMETRYCOLLECTION (MULTIPOLYGON (((-17353464....   

                             centroid  centroidOut  
72  POINT (17226544.434 -2192446.141)          1.0

In [28]:
# Reproject to 4326
countries_4326_1 = reproject_geom_and_centroid(countries6933_1, 4326)
countries_4326_1

COUNTRY   ISO  \
0                                     Afghanistan    AF   
1                           Akrotiri and Dhekelia  None   
2                                         Albania    AL   
3                                         Algeria    DZ   
4                                  American Samoa    AS   
5                                         Andorra    AD   
6                                          Angola    AO   
7                                        Anguilla    AI   
8                                      Antarctica    AQ   
9                             Antigua and Barbuda    AG   
10                                      Argentina    AR   
11                                        Armenia    AM   
12                                          Aruba    AW   
13                                      Australia    AU   
14                                        Austria    AT   
15                                     Azerbaijan    AZ   
16                                        Bahamas    BS   
17                                        Bahrain    BH   
18                                     Bangladesh    BD   
19                                       Barbados    BB   
20                                        Belarus    BY   
21                                        Belgium    BE   
22                                         Belize    BZ   
23                                          Benin    BJ   
24                                        Bermuda    BM   
25                                         Bhutan    BT   
26                                        Bolivia    BO   
27                         Bosnia and Herzegovina    BA   
28                                       Botswana    BW   
29                                  Bouvet Island    BV   
30                                         Brazil    BR   
31                 British Indian Ocean Territory    IO   
32                         British Virgin Islands    VG   
33                                         Brunei    BN   
34                                       Bulgaria    BG   
35                                   Burkina Faso    BF   
36                                        Burundi    BI   
37                                       Cambodia    KH   
38                                       Cameroon    CM   
39                                         Canada    CA   
40                                     Cape Verde    CV   
41                          Caribbean Netherlands  None   
42                                 Cayman Islands    KY   
43                       Central African Republic    CF   
44                                           Chad    TD   
45                                          Chile    CL   
46                                          China    CN   
47                               Christmas Island    CX   
48                                  Cocos Islands    CC   
49                                       Colombia    CO   
50                                        Comoros    KM   
51                                   Cook Islands    CK   
52                                     Costa Rica    CR   
53                                        Croatia    HR   
54                                           Cuba    CU   
55                                         Cyprus    CY   
56                                        Czechia    CZ   
57               Democratic Republic of the Congo    CD   
58                                        Denmark    DK   
59                                       Djibouti    DJ   
60                                       Dominica    DM   
61                             Dominican Republic    DO   
62                                        Ecuador    EC   
63                                          Egypt    EG   
64                                    El Salvador    SV   
65                              Equatorial Guinea    GQ   
66                                        Eritrea    ER   
67                            

In [29]:
# Get the geometry from countries_buffed
# Remember we already ran this: fiji_boundary = countries_buffed[countries_buffed['COUNTRY'] == 'Fiji']
fiji_geo = fiji_boundary['geometry'].iloc[0]

In [30]:
# Get the geometry from fiji_6933
fiji_geo_6933 = fiji_6933['geometry'].iloc[0]

# Replace the geometry value in countries6933 where 'COUNTRY' is 'Fiji'
countries_4326_1.loc[countries_4326_1['COUNTRY'] == 'Fiji', 'geometry'] = fiji_geo

In [31]:
# Should be correct centroid coords now
countries_4326_1[countries_4326_1['COUNTRY'] == 'Fiji']

COUNTRY ISO COUNTRYAFF  Population        formalName  \
72    Fiji  FJ       Fiji      893468  Republic of Fiji   

                                             geometry  \
72  MULTIPOLYGON (((179.99667 -16.53378, 179.98753...   

                      centroid  centroidOut  
72  POINT (178.5388 -17.44728)          1.0

### What a quagmire! As the wikipedia page for the island of Tavueni, Fiji says:
##### "An interesting note is that the island of Taveuni crosses the east–west antimeridian, so the "north-eastern" portion of the island is located at -179 degrees longitude and the south-western part at +179 degrees longitude. This is often an example that causes havoc to GIS software, in which a polygon geometry around the perimeter of the island is incorrectly rendered and wraps around the globe. 

In [ ]:
# # Does everything finally line up??
# # Again, this will take a couple minutes to run
# m = folium.Map(location=[0,0], zoom_start=2)

# # Add only the polygon geometry as GeoJSON (avoiding LineStrings or other non-polygon geometries)
# folium.GeoJson(
#     countries_4326_1[['geometry']],  # Use only the polygons from 'geometry' column
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue'},
#     name="Country Boundaries"
# ).add_to(m)

# # Add only the polygon geometry as GeoJSON (avoiding LineStrings or other non-polygon geometries)
# folium.GeoJson(
#     countries_4326_1[['geometry']],  # Use only the polygons from 'geometry' column
#     style_function=lambda x: {'fillOpacity': 0, 'color': 'blue'},
#     name="Country Boundaries"
# ).add_to(m)

# # OPTIONAL: Save the map
# # m.save('GIS Tutorials/GeoPy/1.Geopandas/activity_files/folium_maps/countries_boundary_check.html')

# # Or show the map
# m

##### Shapefiles can only have one geometry column. So we will split our gdf into two different single geometry gdfs to save as shapefiles.
##### For a version to easily upload back into python/geopandas we'll save it as a parquet

In [32]:
# Split into reproj_geom with main geometry and other attributes except centroid
reproj_geom = countries_4326_1.drop(columns='centroid')

# Split into reproj_centroids with centroid and any non-geometry columns (exclude main geometry)
reproj_centroids = countries_4326_1.drop(columns='geometry')
reproj_centroids = reproj_centroids.set_geometry('centroid')

In [33]:
# # Split into reproj_geom with main geometry and other attributes except centroid
# reproj_geom = countries6933_1.drop(columns='centroid')

# # Split into reproj_centroids with centroid and any non-geometry columns (exclude main geometry)
# reproj_centroids = countries6933_1.drop(columns='geometry')
# reproj_centroids = reproj_centroids.set_geometry('centroid')

In [35]:
reproj_centroids.to_file("GIS Tutorials/Geog-312/geopandas_Files/checkpoint2/countries_w_off_names.shp", driver='ESRI Shapefile')
reproj_geom.to_file("GIS Tutorials/Geog-312/geopandas_Files/checkpoint3/reproj_geom.shp", driver='ESRI Shapefile')

C:\Users\User\AppData\Local\Temp\ipykernel_26516\898618992.py:1: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  reproj_centroids.to_file("GIS Tutorials/Geog-312/Activity_Files/checkpoint2/countries_w_off_names.shp", driver='ESRI Shapefile')
C:\Users\User\miniconda3\envs\geo\Lib\site-packages\pyogrio\raw.py:723: RuntimeWarning: Normalized/laundered field name: 'centroidOut' to 'centroidOu'
  ogr_write(
C:\Users\User\AppData\Local\Temp\ipykernel_26516\898618992.py:2: UserWarning: Column names longer than 10 characters will be truncated when saved to ESRI Shapefile.
  reproj_geom.to_file("GIS Tutorials/Geog-312/Activity_Files/checkpoint3/reproj_geom.shp", driver='ESRI Shapefile')
C:\Users\User\miniconda3\envs\geo\Lib\site-packages\pyogrio\raw.py:723: RuntimeWarning: Normalized/laundered field name: 'centroidOut' to 'centroidOu'
  ogr_write(


In [37]:
# Save as Parquet file
countries_4326_1.to_parquet("GIS Tutorials/Geog-312/geopandas_Files/checkpoint3/reproj.parquet")

# # Or, save as Feather file
# countries6933_1.to_feather("countries_with_centroids.feather")